### ISSUE TITLE
On fedora 38 template selinux preventing using docker with it root dir symlinked to dir in home folder

### ISSUE THREAD

**Issue by Szewcson** [mirage/qubes-mirage-firewall#183]
When tried to follow build instruction of fedora38 based VM I encountered that error:
```
sudo ./build-with-docker.sh
Building Docker image with dependencies..
DEPRECATED: The legacy builder is deprecated and will be removed in a future release.
            Install the buildx component to build images with BuildKit:
            https://docs.docker.com/go/buildx/

Sending build context to Docker daemon  179.2kB
Step 1/13 : FROM debian@sha256:07c6cb2ae86479dcc1942a89b0a1f4049b6e9415f7de327ff641aed58b8e3100
 ---> cb5c02e18207
Step 2/13 : RUN printf "deb [check-valid-until=no] http://snapshot.notset.fr/archive/debian/20230418T024659Z bookworm main" > /etc/apt/sources.list
 ---> Running in 742b578c31d1
/bin/sh: error while loading shared libraries: /lib/x86_64-linux-gnu/libc.so.6: cannot apply additional memory protection after relocation: Permission denied
The command '/bin/sh -c printf "deb [check-valid-until=no] http://snapshot.notset.fr/archive/debian/20230418T024659Z bookworm main" > /etc/apt/sources.list' returned a non-zero code: 127

```

**Comment by palainp** [comment]
Hi @Szewcson, thanks for the report. I'm surprised, to reproduce the issue I just installed a fresh fedora-38 template (just a couple of packages installed (patch, git, etc.)) and I don't suffer a similar issue:
```
$ sudo ./build-with-docker.sh 
Building Docker image with dependencies..
DEPRECATED: The legacy builder is deprecated and will be removed in a future release.
            Install the buildx component to build images with BuildKit:
            https://docs.docker.com/go/buildx/

Sending build context to Docker daemon  176.1kB
Step 1/13 : FROM debian@sha256:07c6cb2ae86479dcc1942a89b0a1f4049b6e9415f7de327ff641aed58b8e3100
docker.io/library/debian@sha256:07c6cb2ae86479dcc1942a89b0a1f4049b6e9415f7de327ff641aed58b8e3100: Pulling from library/debian
143db7d08421: Pull complete 
Digest: sha256:07c6cb2ae86479dcc1942a89b0a1f4049b6e9415f7de327ff641aed58b8e3100
Status: Downloaded newer image for debian@sha256:07c6cb2ae86479dcc1942a89b0a1f4049b6e9415f7de327ff641aed58b8e3100
 ---> cb5c02e18207
Step 2/13 : RUN printf "deb [check-valid-until=no] http://snapshot.notset.fr/archive/debian/20230418T024659Z bookworm main" > /etc/apt/sources.list
 ---> Running in 750e570c58c5
Removing intermediate container 750e570c58c5
 ---> e046a90189e3
Step 3/13 : RUN apt update && apt install --no-install-recommends --no-install-suggests -y wget ca-certificates git patch unzip bzip2 make gcc g++ libc-dev
 ---> Running in 71bc2dec88f1

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:2 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]
...
cat /dev/null | gzip -n > _build/mirage-firewall/initramfs
tar cjf mirage-firewall.tar.bz2 -C _build --mtime=./build-with-docker.sh mirage-firewall
sha256sum mirage-firewall.tar.bz2 > mirage-firewall.sha256
SHA2 of build:   8ae5314edf5b863b788c4b873e27bc4b206a2ff7ef1051c4c62ae41584ed3e14  ./dist/qubes-firewall.xen
SHA2 last known: 8ae5314edf5b863b788c4b873e27bc4b206a2ff7ef1051c4c62ae41584ed3e14
(hashes should match for released versions)
```
Maybe it could be related to a template change (from f-37 to f-38) of your appVM and docker has changed some cache files in the meantime? Note that I'm not yet with Qubes 4.2, still with 4.1 and I don't know if it could matters here.

**Comment by Szewcson** [comment]
I'm on 4.2.

**Comment by palainp** [comment]
Thinking twice it could be due to 4.1 vs 4.2 as Qubes onboard the AppVM linux kernel.
When 4.2 will be released I'll be able to investigate further. Thanks again for the report!

**Comment by Szewcson** [comment]
From Qubes OS 4.2.0 release notes:

- SELinux support in Fedora templates ([#4239](https://github.com/QubesOS/qubes-issues/issues/4239))

**Comment by palainp** [comment]
Hi, I now have a functional Qubes 4.2-rc3 laptop and while reproducing this I found that adding the `Z` label to the build volume made it work as before/expected.

So I updated `build-with-docker.sh` as:
```bash
#!/bin/sh
set -eu
echo Building Docker image with dependencies..
docker build -t qubes-mirage-firewall .
echo Building Firewall...
docker run --rm -i -v `pwd`:/tmp/orb-build:Z qubes-mirage-firewall
echo "SHA2 of build:   $(sha256sum ./dist/qubes-firewall.xen)"
echo "SHA2 last known: 8ae5314edf5b863b788c4b873e27bc4b206a2ff7ef1051c4c62ae41584ed3e14"
echo "(hashes should match for released versions)"
```

As I'm not a Docker expert, I understand the `man docker-run` documentation (
```
       To change a label in the container context, you can add either of two suffixes :z or :Z to the volume mount. These suffixes tell
       Docker to relabel file objects on the shared volumes. The z option tells Docker that two containers share the volume content. As
       a  result,  Docker  labels the content with a shared content label. Shared volume labels allow all containers to read/write con‐
       tent.  The Z option tells Docker to label the content with a private unshared label.  Only the current container can use a  pri‐
       vate volume.
```
) as : the suffixes z/Z indicate that a shared label should be added to the volume files and the capital one makes it accessible to only one specific container. Since we're only using docker for reproducibility purposes, I think it's possible to make this change without security issues. Any thoughts?

**Comment by palainp** [comment]
Hi @Szewcson , if you want to try it out, I made a branch some times ago with a fix at https://github.com/palainp/qubes-mirage-firewall/tree/fix-docker-build.

**Comment by Szewcson** [comment]
Still the same

**Comment by palainp** [comment]
Yeah, I got mixed up in the docker cache and didn't realize the build directive wasn't being re-runned, sorry about that.

I've updated the branch with a new patch (an additional command in README.md) to update the `/home/user/docker` directory SELinux context (`sudo chcon -Rt container_file_t /home/user/docker` in the docker installation notes). It ended to with a matching shasum in a new dispVM :)

**Comment by palainp** [comment]
Great! For the hashum mismatch see #184. I'm working on it.

**Comment by palainp** [comment]
Thanks, fixed in #185.
